### ISSUE TITLE
Make random subsequence sampling reproducible across seeded runs

### ISSUE THREAD

**Issue by florian6973** [mmcdermott/meds-torch-data#116]
## Summary

Normal `MEDSPytorchDataset[idx]` access does not connect stochastic subsequence sampling to an experiment seed. Consequently, `random` and `balanced_random` sampling cannot currently be reproduced across identical seeded runs.

The stochastic behavior itself is desirable; the missing reproducibility contract is the issue.

## Confirmed behavior

`MEDSPytorchDataset.__getitem__` calls:

```python
return self._seeded_getitem(idx)
```

without a seed. That `None` is forwarded through `process_dynamic_data(..., rng=seed)` to `resolve_rng(None)`, which returns a fresh `numpy.random.default_rng()`.

A local reproduction reset both `torch.manual_seed(42)` and `numpy.random.seed(42)` before each draw. Two calls through `resolve_rng(None)` produced different sequences, confirming that global Lightning/PyTorch seeding does not control this RNG.

The README currently documents normal random access as non-deterministic and demonstrates determinism only through the internal `_seeded_getitem`. This is therefore a reproducibility gap in the public training path, not a regression against an existing deterministic promise.

## Desired contract

When an explicit dataset sampling seed is configured:

```text
same base seed + same epoch + same dataset index
    => same sampled subsequence
```

Different epochs should be able to produce different subsequences, and the result for a given tuple should not depend on DataLoader worker assignment or scheduling.

Unseeded behavior should remain available for backward compatibility.

## Proposed design

1. Add an optional explicit configuration value such as `subsequence_sampling_seed: int | None = None`.
2. Add `MEDSPytorchDataset.set_epoch(epoch)`.
3. Derive a generator using stable NumPy seed composition, for example:

```python
np.random.default_rng(np.random.SeedSequence([base_seed, epoch, idx]))
```

4. Keep epoch state visible to persistent DataLoader worker copies.
5. Expose epoch propagation through the Lightning datamodule.
6. Seed the training DataLoader's dedicated `torch.Generator` from the same explicit seed so shuffle order is isolated from unrelated global RNG consumption.

Do not use Python's built-in `hash()`, which is unsuitable as a persistent reproducibility contract.

## Acceptance tests

- same seed, epoch, and index gives the same subsequence
- changing seed or epoch can change the subsequence
- deterministic strategies remain unchanged
- results match with `num_workers=0` and `num_workers>0`
- persistent workers observe epoch updates
- two complete seeded runs produce identical sample and batch order
- the unseeded default remains stochastic
